


## 1. Introduction

The objectives of this lab are:

    1. To familiarize you with how to create dense neural networks using Keras.
    2. To familiarize you with how to encode input and output vectors for neural networks.
    3. To give you some insight into how hyperparameters like learning rate and momentum affect training.
    
To save time we will train each experiment only for 50 epochs. This will lead to less than optimal results but is enough for you to make observations.

**HINT: YOU CAN HIT SHIFT-ENTER TO RUN EACH CELL. NOTE THAT IF A CELL IS DEPENDENT ON A PREVIOUS CELL, YOU WILL NEED TO RUN THE PREVIOUS CELL(S) FIRST **


## 2. The Irises Dataset

We will now work again on the Irises Dataset, which we used in Lab 3, for classifying iris flowers into one of three possible types. As before we will consider four factors:

    1. Sepal length in cm
    2. Sepal width in cm
    3. Petal length in cm
    4. Petal width in cm

In this dataset there are 150 sample points. The code below loads the dataset and prints the first 10 rows so we have an idea of what it looks like.

In [1]:
from sklearn.datasets import load_iris
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

iris = load_iris()

print("First 10 rows of data:")
print(iris.data[:10])

First 10 rows of data:
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]]


### 2.2 Scaling the Data

We make use of the MinMaxScaler to scale the inputs to between 0 and 1.  The code below does this and prints the first 10 rows again, to show us the difference.

In the next section we will investigate what happens if we use unscaled data.

In [2]:
scaler = MinMaxScaler()
scaler.fit(iris.data)
X = scaler.transform(iris.data)

print("First 10 rows of SCALED data.")
print(X[:10])

First 10 rows of SCALED data.
[[0.22222222 0.625      0.06779661 0.04166667]
 [0.16666667 0.41666667 0.06779661 0.04166667]
 [0.11111111 0.5        0.05084746 0.04166667]
 [0.08333333 0.45833333 0.08474576 0.04166667]
 [0.19444444 0.66666667 0.06779661 0.04166667]
 [0.30555556 0.79166667 0.11864407 0.125     ]
 [0.08333333 0.58333333 0.06779661 0.08333333]
 [0.19444444 0.58333333 0.08474576 0.04166667]
 [0.02777778 0.375      0.06779661 0.04166667]
 [0.16666667 0.45833333 0.08474576 0.        ]]


### 2.3 Encoding the Targets

In Lab 3 we saw that the target values (type of iris flower) is a vector from 0 to 2. We can see the 150 labels below:


In [3]:
print(iris.target)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


We can use this to train the neural network, but we will use "one-hot" encoding, where we have a vector of _n_ integers consisting of 0's and 1's.  The table below shows how one-hot encoding works:

|   Value    |    One-Hot Encoding    |
|:----------:|:----------------------:|
| 0 | \[1 0 0\] |
| 1 | \[0 1 0\] |
| 2 | \[0 0 1\] |

Keras provides the to_categorical function to create one-hot vectors:



In [4]:
from keras.utils import to_categorical

Y = to_categorical(y = iris.target, num_classes = 3)
print(Y)

[[1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0.

Now let's split the data into training and testing data:



In [5]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, 
                                                    random_state = 1)


### 2.4 Building our Neural Network

Let's now begin building a simple neural network with a single hidden layer, using the Stochastic Gradient Descent (SGD) optimizer, ReLu transfer functions for the hidden layer and softmax for the output layer.

The code to do this is shown below:

In [10]:
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import SGD

# Create the neural network
nn = Sequential()
nn.add(Dense(100, input_shape = (4, ), activation = 'relu'))
nn.add(Dense(3, activation = 'softmax'))

# Create our optimizer
sgd = SGD(lr = 0.1)

# 'Compile' the network to associate it with a loss function,
# an optimizer, and what metrics we want to track
nn.compile(loss='categorical_crossentropy', optimizer=sgd,
          metrics = 'accuracy')

### 2.5 Training the Neural Network

As is usually the case, we can call the "fit" method to train the neural network for 50 epochs. We will shuffle the training data between epochs, and provide validation data.

In [16]:
nn.fit(X_train, Y_train, shuffle = True, epochs = 50, 
      validation_data = (X_test, Y_test))

Epoch 1/50
4/4 [==============================] - 0s 31ms/step - loss: 0.1076 - accuracy: 0.9667 - val_loss: 0.1281 - val_accuracy: 0.9667
Epoch 2/50
4/4 [==============================] - 0s 16ms/step - loss: 0.1093 - accuracy: 0.9667 - val_loss: 0.1108 - val_accuracy: 0.9667
Epoch 3/50
4/4 [==============================] - 0s 12ms/step - loss: 0.1029 - accuracy: 0.9667 - val_loss: 0.0949 - val_accuracy: 0.9667
Epoch 4/50
4/4 [==============================] - 0s 12ms/step - loss: 0.1054 - accuracy: 0.9667 - val_loss: 0.0970 - val_accuracy: 0.9667
Epoch 5/50
4/4 [==============================] - 0s 21ms/step - loss: 0.1026 - accuracy: 0.9750 - val_loss: 0.1063 - val_accuracy: 0.9667
Epoch 6/50
4/4 [==============================] - 0s 18ms/step - loss: 0.1036 - accuracy: 0.9667 - val_loss: 0.1010 - val_accuracy: 0.9667
Epoch 7/50
4/4 [==============================] - 0s 23ms/step - loss: 0.1029 - accuracy: 0.9667 - val_loss: 0.0773 - val_accuracy: 0.9667
Epoch 8/50
4/4 [===========

In [30]:
# # Create the neural network, playing with learning rates
# nn = Sequential()
# nn.add(Dense(100, input_shape = (4, ), activation = 'relu'))
# nn.add(Dense(3, activation = 'softmax'))

# # Create our optimizer
# sgd = SGD(lr = 100000)

# # 'Compile' the network to associate it with a loss function,
# # an optimizer, and what metrics we want to track
# nn.compile(loss='categorical_crossentropy', optimizer=sgd,
#           metrics = 'accuracy')

# nn.fit(X_train, Y_train, shuffle = True, epochs = 50,
#       validation_data = (X_test, Y_test))

Epoch 1/50
4/4 [==============================] - 1s 95ms/step - loss: 112019258510684991782912.0000 - accuracy: 0.4167 - val_loss: 17600.1367 - val_accuracy: 0.3667
Epoch 2/50
4/4 [==============================] - 0s 23ms/step - loss: 1600028288.0000 - accuracy: 0.3833 - val_loss: 36532.6367 - val_accuracy: 0.2000
Epoch 3/50
4/4 [==============================] - 0s 22ms/step - loss: 29165.0254 - accuracy: 0.3917 - val_loss: 26181797675008.0000 - val_accuracy: 0.4333
Epoch 4/50
4/4 [==============================] - 0s 19ms/step - loss: 9282576711680.0000 - accuracy: 0.3750 - val_loss: 41220.1445 - val_accuracy: 0.2000
Epoch 5/50
4/4 [==============================] - 0s 27ms/step - loss: 39815.1133 - accuracy: 0.2833 - val_loss: 26694.3691 - val_accuracy: 0.3667
Epoch 6/50
4/4 [==============================] - 0s 16ms/step - loss: 31279.7461 - accuracy: 0.3583 - val_loss: 28303.4785 - val_accuracy: 0.2000
Epoch 7/50
4/4 [==============================] - 0s 21ms/step - loss: 26404.

In [ ]:
# # Create the neural network - Testing Momentum
# nn = Sequential()
# nn.add(Dense(100, input_shape = (4, ), activation = 'relu'))
# nn.add(Dense(3, activation = 'softmax'))

# # Create our optimizer
# sgd = SGD(lr = 0.1,momentum=1)

# # 'Compile' the network to associate it with a loss function,
# # an optimizer, and what metrics we want to track
# nn.compile(loss='categorical_crossentropy', optimizer=sgd,
#           metrics = 'accuracy')

# nn.fit(X_train, Y_train, shuffle = True, epochs = 50,
#       validation_data = (X_test, Y_test))

---
#### Question 1

Run the code above. Do you see evidence of underfitting? Overfitting? Justify your answers. ***(4 MARKS)***


**Answer: The model is Overfitting as the validation loss is not minimum for the last epoch. Val_loss for epoch 49 was 0.0884  which is lower than 0.0905, the val_loss of epcoh 50.**

_(For TA) Marks awarded: ____ / 4_

---

#### Question 2a

Consult the documentation for the SGD optimizer [here](https://keras.io/api/optimizers/sgd/). What does the lr parameter do? ***(1 MARK)***

**Answer: lr is learning rate, The learning rate is a hyperparameter that controls how much to change the model in response to the estimated error each time the model weights are updated.**

#### Question 2b

The documentation states that the momentum parameter "accelerates gradient descent in the relevant direction and dampens oscillations". Using Google or other means, illustrate what this means. ***(2 MARKS)***

**Answer: Essentially, when using momentum, we push a ball down a hill. The ball accumulates momentum as it rolls downhill, becoming faster and faster on the way. The momentum parameter increases the velocity if we are moving towards lowest error rate and decreases if we are moving away from local minima, which helps us to reach local minima i.e. lower error rate. Oftenly with large learning rate, we could face the issues of oscillating away from local minima, this momentum parameter helps us avoid that and dampens the oscillations, hence makes it possible to achieve lowest error rate.**

_(For TA) Marks awarded: ____ / 3_

----

#### Question 3a

We will now play with the lr parameter. Adjust the lr parameter to the following values and record the final training and validation accuracies in the respective columns. Also observe the sequence of accuracies over the training period, and place your observation in the "remarks" column, e.g. "Progresses steadily", "some oscillation" etc.***(3 MARKS)***

**Answer: Fill the table below **

|  lr    | Training Acc. | Validation Acc. |      Remarks      |
|:------:|---------------|-----------------|-------------------|
|0.01    |     0.6917          |    0.5667          |   Progresses steadily                |
|0.1     |    0.9750           |   0.9667           |   Progresses steadily                |
|1.0     |     0.9          |       1.0          |      Progresses              |
|10.0    |     0.33          |       0.37          |    some oscillation               |
|100     |     0.325          |       0.433          |    some oscillation               |
|1000    |      0.3         |        0.433         |     some oscillation              |
|10000   |      0.77         |       0.57          |     some oscillation              |
| 100000 |       0.28        |      0.433           |   some oscillation                |


#### Question 3b

Based on your observations above, comment on the effect of small and very large learning rates on the learning. ***(2 MARKS)***

**Answer: Smaller learning rate showed Slower progress towards increase in accuracy, whereas moderate learning rates give very good accuracy. Large learning rates forced oscillations as we deviate from local minima, hence the lower accuracies and oscillations observed.**

_(For TA) Marks awarded: ____ / 5_

### 2.5 Using Momentum

We will now experiment with the momentum term. To do this:

    1. Change the learning rate to 0.1.
    2. Set the momentum to 0.1. Note: Do not use the Nesterov parameter - Leave it as False.
    
Run your neural network.

---

#### Question 4a

Keeping the learning rate at 0.1, complete the table below using the momentum values shown. Again record any observations in the "Remarks" column. ***(3 MARKS)***

**Answer: Fill the table below**

| momentum | Training Acc. | Validation Acc. |      Remarks      |
|:--------:|---------------|-----------------|-------------------|
|0.001     |   0.9333            |     0.9333            |      steady progress             |
|0.01      |    0.925           |       0.9667           |      steady progress             |
|0.1       |    0.925           |      0.97           |      steady progress              |
|1.0       |    0.925            |           1      |      steady progress              |

#### Question 4b

Based on your observations above, does the momentum term help in learning? ***(2 MARKS)***

**Answer: Yes, The Validation accuracy has gradually improved with the increase in Momentum. Indicating betting learning as compared to default momentum value.**

_(For TA) Marks awarded: ____ / 5_

---

### 2.6 Using Raw Unscaled Data

We begin by using unscaled X and Y data. The code below will create 120 training samples and 30 testing samples (20% of the total of 150 samples):

In [35]:
X_unscaled = iris.data
Y_raw = iris.target
X_utrain, X_utest, Y_utrain, Y_utest = train_test_split(X_unscaled, Y,
                                                        test_size = 0.2,
                                                        random_state = 1)

---

#### Question 5

Create a new neural network called "nn2" below using a single hidden layer of 100 neurons. Train using the data in X_utrain, X_utest and validate with Y_utrain and Y_utest. Again use the SGD optimizer with a learning rate of 0.1 and no momentum, and train for 50 epochs. ***(3 marks)***

In [65]:
"""
Enter your code for Question 5 below. To TA: 3 marks for code.
"""
# Create the neural network
nn2 = Sequential()
nn2.add(Dense(100, input_shape = (4, ), activation = 'relu'))
nn2.add(Dense(3, activation = 'softmax'))

# Create our optimizer
sgd2 = SGD(lr = 0.1)

# 'Compile' the network to associate it with a loss function,
# an optimizer, and what metrics we want to track
nn2.compile(loss='categorical_crossentropy', optimizer=sgd2,
          metrics = 'accuracy')

nn2.fit(X_utrain, Y_utrain, shuffle = True, epochs = 50,
      validation_data = (X_utest, Y_utest))



Epoch 1/50
4/4 [==============================] - 1s 139ms/step - loss: 1.2467 - accuracy: 0.4000 - val_loss: 2.1555 - val_accuracy: 0.5667
Epoch 2/50
4/4 [==============================] - 0s 14ms/step - loss: 1.1460 - accuracy: 0.5167 - val_loss: 0.9508 - val_accuracy: 0.5667
Epoch 3/50
4/4 [==============================] - 0s 11ms/step - loss: 0.6736 - accuracy: 0.7000 - val_loss: 0.6021 - val_accuracy: 0.6667
Epoch 4/50
4/4 [==============================] - 0s 19ms/step - loss: 0.5935 - accuracy: 0.7583 - val_loss: 0.6589 - val_accuracy: 0.5667
Epoch 5/50
4/4 [==============================] - 0s 29ms/step - loss: 0.4915 - accuracy: 0.8583 - val_loss: 0.5936 - val_accuracy: 0.6000
Epoch 6/50
4/4 [==============================] - 0s 18ms/step - loss: 0.5143 - accuracy: 0.7250 - val_loss: 0.7527 - val_accuracy: 0.5667
Epoch 7/50
4/4 [==============================] - 0s 22ms/step - loss: 0.4787 - accuracy: 0.7667 - val_loss: 0.5907 - val_accuracy: 0.6000
Epoch 8/50
4/4 [==========

**(Question 5 continues)**

Observe the training and validation error. Does not scaling the input affect the training? Why do you think this is so? What is the advantage of scaling? ***(5 MARKS)***

**Answer:Both the Training and validation errors for unscaled data are greater than what we have for scaled data. The accuracies are low as well. In the case of Neural Networks Algorithms, feature scaling benefits optimization by: It makes the training faster. It prevents the optimization from getting stuck in local optima. It gives a better error surface shape.**

_(For TA) Marks awarded: ____ / 8_

---